src/cc/frontends/clang/b_frontend_action.cc

```diff
@@ -487,6 +488,52 @@ bool ProbeVisitor::VisitMemberExpr(MemberExpr *E) {
   rewriter_.ReplaceText(expansionRange(SourceRange(member, E->getLocEnd())), post);
   return true;
 }
+bool ProbeVisitor::VisitArraySubscriptExpr(ArraySubscriptExpr *E) {
+  if (memb_visited_.find(E) != memb_visited_.end()) return true;
+  if (!ProbeChecker(E, ptregs_, track_helpers_).needs_probe()) {
+    return true;
+  }
+
+  // parent expr has addrof, skip the rewrite, set is_addrof_ to flase so
+  // it won't affect next level of indirect address
+  if (is_addrof_) {
+    return true;
+  }
+
+  Expr *base = E->getBase();
+  Expr *idx = E->getIdx();
+  memb_visited_.insert(E);
+  // If the base of the array subscript is a member dereference, we'll rewrite
+  // both at the same time.
+  if (base->IgnoreParenCasts()->getStmtClass() == Stmt::MemberExprClass)
+    memb_visited_.insert(base->IgnoreParenCasts());
+
+  string pre, lbracket, rbracket;
+  LangOptions opts;
+  SourceLocation lbracket_start, lbracket_end;
+  SourceRange lbracket_range;
+  pre = "({ typeof(" + E->getType().getAsString() + ") _val; __builtin_memset(&_val, 0, sizeof(_val));";
+  pre += " bpf_probe_read(&_val, sizeof(_val), (u64)(";
+  if (base->IgnoreParenCasts()->getStmtClass() == Stmt::MemberExprClass)
+    pre += "&";
```

Comment:
I already `memb_visited_.insert(base->IgnoreParenCasts());` above, so that shouldn't be needed, no?

Reply:
Looks like marking `memb_visited_` for `base` is not enough. After rewriting, the clang seems generating a new expression for the `base` and base pointers before and after transformation won't match. In the above, I am using `addrof_stmt_ = E` which I suspect may not really work since after rewrite, the expression is not a subscript expr any more.

Reply:
I just checked to be sure: except for `test_probe_read_array_accesses5`, all the new test cases translate into a single call to `bpf_probe_read`. If `memb_visited_.insert(base->IgnoreParenCasts())` wasn't enough and Clang was changing the base pointers, we would have several calls to `bpf_probe_read`, as least for the 3 first tests, right?

Are you sure the base pointer can change after a rewrite? I don't think Clang updates its AST after each rewriter; as far as I understand, that's why we need `expansionRange`s.

Am I missing something? Do you have another test case I should be checking?

Reply:
I am using facebook internal applications for testing to make sure they still work.
i printed out the address of the array subscript expr, member expr and expr address after `RecursiveASTVisitor<ProbeVisitor>::TraverseStmt`. You are right, the base pointer and various expression address seems still valid.

Further checking found all the offending unnecessary bpf_probe_read rewrite has the
pattern like below:
```
   sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32[0]
```
Baiscally
```
  base->member.field[subscript]
```
The patch did
```
bpf_probe_read(..., &base->member.field[subscript], ...);
```
and later `base->member` is rewritten as
```
bpf_probe_read(..., &(... bpf_probe_read(..., &base->member, ...) ...).field[subscript], ...);
```
And `({...; _val}).field is illegal, for example, it could be an anonymous struct/union in many cases.

the following hack can be used to reproduce the issue
```
diff --git a/tools/tcptop.py b/tools/tcptop.py
index 7d2babb..51241ca 100755
--- a/tools/tcptop.py
+++ b/tools/tcptop.py
@@ -118,7 +118,7 @@ int kprobe__tcp_sendmsg(struct pt_regs *ctx, struct sock *sk,
     } else if (family == AF_INET6) {
         struct ipv6_key_t ipv6_key = {.pid = pid};
 
-        ipv6_key.saddr0 = *(u64 *)&sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32[0];
+        ipv6_key.saddr0 = sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32[0];
         ipv6_key.saddr1 = *(u64 *)&sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32[2];
         ipv6_key.daddr0 = *(u64 *)&sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32[0];
         ipv6_key.daddr1 = *(u64 *)&sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32[2];
```

Reply:
Another example,
```
mm->rss_stat.count[MM_ANONPAGES].counter
```
```
struct mm_rss_stat {
        atomic_long_t count[NR_MM_COUNTERS];
};
```
linux seems not liking bpf_probe_read for `mm->rss_stat`.

Reply:
Thanks for the details and the test cases @yonghong-song! I always forget that not all member expressions are dereferences :-/

I think I have a fix for this. It adds a bit of code, but I might be able to refactor. I'll try to double check and update the branch tonight.

Reply:
@pchaigno I will be traveling in the next three weeks and may not response in time. @palmtenor  will help do review as well. Thanks!

Reply:
No problem, there's no hurry.

I've updated the branch; it should work with member expressions now. I've added two test cases (7 & 8) for this.

Basically, instead of checking only if the base is a member expression, we now check if the base is a member *dereference* (`a->b`) followed by 0 to n member accesses (`a->b.c.d`). If it is, we set `is_addrof_`.